models/generators.py

```diff
@@ -1,67 +1,91 @@
 import tensorflow as tf
 from tensorflow.keras import Model
-from tensorflow.keras.layers import (Activation, BatchNormalization, Conv2D,
-                                     Dense, Reshape, UpSampling2D, Conv2DTranspose,
-                                     ReLU)
+from tensorflow.keras.layers import (BatchNormalization, Conv2D, LeakyReLU,
+                                     Dense, Reshape, Conv2DTranspose,
+                                     ReLU, Flatten)
 from tensorflow.keras.activations import tanh
-
-from utils.utils import product_list
+from utils.utils import sample_normal
 
 
 class Generator(Model):
     """ The definition for a network which
         fabricates images from a noisy distribution.
     """
-    def __init__(self, img_size, num_latent_dims, kernel_size, num_filters,
-                 reshape_dims):
+    def __init__(self, img_size, num_latent_dims, lr, conditional_emb_size):
         """ Initialise a Generator instance.
             TODO: Deal with this parameters and make it more logical
-                Arguments:
+                Arguments:'
                 img_size : tuple of ints
                     Size of images. E.g. (1, 32, 32) or (3, 64, 64).
                 num_latent_dims : int
                     Dimensionality of latent input.
-                kernel_size : tuple
-                    (height, width)
-                reshape_dims : tuple or list
-                    [91, 125, 128]
+                lr: float
+                    Learning rate
+                conditional_emb_size: Tensor
+                    text embedding. Shape (batch_size, feature_size, embedding_size)
         """
         super().__init__()
         self.img_size = img_size
-        num_output_channels = self.img_size[0]
         self.num_latent_dims = num_latent_dims
-        self.xent_loss_fn = tf.keras.losses.BinaryCrossentropy(from_logits=False)
-        self.optimiser = tf.keras.optimizers.Adam(1e-4)
-        # TODO: Add correct layers as given by the paper
-        self.dense1 = Dense(units=16*16*32, activation='relu') # NOTE make dynamic
-        self.reshape_layer = Reshape([16, 16, 32])
-        self.upsample1 = UpSampling2D()
-        self.conv1 = Conv2D(filters=num_filters, kernel_size=kernel_size, activation='relu', padding='same')
-        self.upsample2 = UpSampling2D()
-        self.conv2 = Conv2D(filters=num_output_channels, kernel_size=kernel_size, activation='tanh', padding='same')
+
+        # Optimiser
+        self.optimiser = tf.keras.optimizers.Adam(lr, beta_1=0.5)
+
+        # Weight Initialisation Parameters
+        self.w_init = tf.random_normal_initializer(stddev=0.02)
+        self.bn_init = tf.random_normal_initializer(1., 0.02)
```

Comment:
these should be moved out of here, they are reused in the discriminator too. Either live in stackGAN or a utils file maybe, then passed as arguments?

Reply:
Ah right. Yeah I'll make that change

Reply:
I'm not going to fiddle the discriminator on this PR. I'll do that right after though